modules/account-lib/src/coin/hbar/tokenTransferBuilder.ts

```diff
@@ -0,0 +1,115 @@
+import { BaseCoin as CoinConfig } from '@bitgo/statics';
+import * as Long from 'long';
+import { TokenId } from '@hashgraph/sdk';
+import * as proto from '@hashgraph/proto';
+import { BuildTransactionError, InvalidParameterValueError } from '@bitgo/sdk-core';
+import { TransferBuilder, Recipient } from './transferBuilder';
+import { Transaction } from './transaction';
+import { getHbarTokenFromTokenName, getHbarTokenFromId, stringifyAccountId, stringifyTokenId } from './utils';
+import { BigNumber } from 'bignumber.js';
+
+export class TokenTransferBuilder extends TransferBuilder {
+  constructor(_coinConfig: Readonly<CoinConfig>) {
+    super(_coinConfig);
+  }
+
+  /** @inheritdoc */
+  protected async buildImplementation(): Promise<Transaction> {
+    this._txBodyData.tokenTransfers = this.buildTokenTransferData(); // set to list by the contract
+    return await super.buildImplementation();
+  }
+
+  private buildTokenTransferData(): proto.ITokenTransferList[] {
+    const tokenTransfers: proto.ITokenTransferList[] = [];
+    const tokenTransferAmount: Record<string, BigNumber> = {}; // total send amount for each token
+    const tokenTransferData: Record<string, proto.IAccountAmount[]> = {}; // accountAmount for each token
+
+    this._recipients.map((recipient) => {
+      const tokenName = recipient.tokenName!;
+      const tokenAmount = tokenTransferAmount[tokenName] || new BigNumber(0);
+      tokenTransferAmount[tokenName] = tokenAmount.plus(recipient.amount);
+      if (!(tokenName in tokenTransferData)) {
+        tokenTransferData[tokenName] = [
+          {
+            accountID: this.buildAccountData(this._source.address),
+            amount: Long.fromInt(0),
+          },
+        ];
+      }
+      tokenTransferData[tokenName].push(
+        { accountID: this.buildAccountData(recipient.address), amount: Long.fromString(recipient.amount) }, // recipient
+      );
+    });
+
+    Object.entries(tokenTransferData).forEach(([tokenName, tokenData]) => {
+      tokenData[0].amount = Long.fromString(tokenTransferAmount[tokenName].toString()).negate(); // update sender send amount
+      const tokenId = getHbarTokenFromTokenName(tokenName)!.tokenId!;
+      tokenTransfers.push({ token: this.buildTokenId(tokenId), transfers: tokenData });
+    });
+
+    return tokenTransfers;
+  }
+
+  private buildTokenId(tokenId: string): proto.TokenID {
+    const tokenData = TokenId.fromString(tokenId);
+    return new proto.TokenID({
+      tokenNum: tokenData.num,
+      realmNum: tokenData.realm,
+      shardNum: tokenData.shard,
+    });
+  }
+
+  /** @inheritdoc */
+  initBuilder(tx: Transaction): void {
+    super.initBuilder(tx);
+    const transferData = tx.txBody.cryptoTransfer;
+    if (transferData && transferData.tokenTransfers && transferData.tokenTransfers.length > 0) {
```

Comment:
You can avoid some redundancy here and replace this with `if (isTokenTransfer(transferData))`